src/extensions/arith.rs

```diff
@@ -709,39 +915,96 @@ pub enum Arith<T: ExtParam> {
     Geq(Expr<T>, Expr<T>),
 }
 
+/// Wrapper around [`ArithInner`] that ensures that the expression is valid.
+/// See [`ArithInner`] for more details.
+///
+/// Note that the library allows construction of unchecked [`Expr], but
+/// [`Arith`] is always checked.
+#[derive(Eq, PartialEq, Ord, PartialOrd, Hash, Clone)]
+pub struct Arith<T: ExtParam> {
+    /// The underlying expression
+    expr: ArithInner<T>,
+}
+
+impl<T: ExtParam> Arith<T> {
+    /// Create a new Arith expression. This is the only constructor
+    pub fn new(expr: ArithInner<T>) -> Result<Self, TypeError> {
+        {
+            // Borrow checker scope
+            let (a, b) = match &expr {
+                ArithInner::Eq(ref a, ref b)
+                | ArithInner::Lt(ref a, ref b)
+                | ArithInner::Leq(ref a, ref b)
+                | ArithInner::Gt(ref a, ref b)
+                | ArithInner::Geq(ref a, ref b) => (a, b),
+            };
+            let mut iter = a.iter_terminals();
+            if let Some(ExprInner::PriceOracle1W(_, _)) = iter.next() {
+                return Err(TypeError::PriceOracle1WFirst);
+            }
+            // Note iter here has consumed the first element
+            if iter.any(|x| {
+                if let ExprInner::PriceOracle1(_, _) = x {
+                    true
+                } else {
+                    false
+                }
+            }) {
+                return Err(TypeError::PriceOracle1Missing);
+            }
+            // All the elements in b should be PriceOracle1W
+            if b.iter_terminals().any(|x| {
+                if let ExprInner::PriceOracle1(_, _) = x {
+                    true
+                } else {
+                    false
+                }
+            }) {
+                return Err(TypeError::PriceOracle1Missing);
+            }
+        }
+        Ok(Arith { expr })
```

Comment:
In 5e533b71c8904f8ad2e0f6ec7ec50f33e3ecd13a:

nit: I think this function could be clearer by (a) using `.chain` to link the two iterators so you can do the `PriceOracle1Missing` check on both of them at once; (b) using the `matches!` macro which we have now since bumping MSRV to 1.48 (though maybe we haven't done that here..)

```